File: yarl-java/src/test/java/hr/tjakopan/yarl/retry/WaitAndRetryForeverHandleExceptionTest.java
package hr.tjakopan.yarl.retry;

import hr.tjakopan.yarl.Context;
import hr.tjakopan.yarl.test.helpers.PolicyUtils;
import org.junit.Test;

import java.time.Duration;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicReference;
import java.util.stream.Collectors;

import static hr.tjakopan.yarl.Functions.fromConsumer;
import static hr.tjakopan.yarl.Functions.fromConsumer4;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;

public class WaitAndRetryForeverHandleExceptionTest {
  @Test
  public void shouldThrowWhenSleepDurationsIsNull() {
    //noinspection ConstantConditions
    assertThatExceptionOfType(NullPointerException.class).isThrownBy(() ->
      RetryPolicy.<Void>builder()
        .handle(ArithmeticException.class)
        .waitAndRetryForever(null,
          fromConsumer4(r -> d -> (i, c) -> {
          })))
      .withMessageContaining("sleepDurationProvider");
  }

  @Test
  public void shouldThrowWhenOnRetryActionIsNull() {
    //noinspection ConstantConditions
    assertThatExceptionOfType(NullPointerException.class).isThrownBy(() ->
      RetryPolicy.<Void>builder()
        .handle(ArithmeticException.class)
        .waitAndRetryForever((i, r, c) -> Duration.ZERO, null))
      .withMessageContaining("onRetry");
  }

  @Test
  public void shouldNotThrowRegardlessOfHowManyTimesTheSpecifiedExceptionIsRaised() {
    final var policy = RetryPolicy.<Void>builder()
      .handle(ArithmeticException.class)
      .waitAndRetryForever((i, r, c) -> Duration.ZERO);

    PolicyUtils.raiseExceptions(policy, 3, i -> new ArithmeticException());
  }

  @Test
  public void shouldNotThrowRegardlessOfHowManyTimesOneOfTheSpecifiedExceptionIsRaised() {
    final var policy = RetryPolicy.<Void>builder()
      .handle(ArithmeticException.class)
      .handle(IllegalArgumentException.class)
      .waitAndRetryForever((i, r, c) -> Duration.ZERO);

    PolicyUtils.raiseExceptions(policy, 3, i -> new IllegalArgumentException());
  }

  @Test
  public void shouldThrowWhenExceptionThrownIsNotTheSpecifiedExceptionType() {
    final var policy = RetryPolicy.<Void>builder()
      .handle(ArithmeticException.class)
      .waitAndRetryForever((i, r, c) -> Duration.ZERO);

    assertThatExceptionOfType(NullPointerException.class).isThrownBy(() ->
      PolicyUtils.raiseExceptions(policy, 1, i -> new NullPointerException()));
  }

  @Test
  public void shouldThrowWhenExceptionThrownIsNotOneOfTheSpecifiedExceptionTypes() {
    final var policy = RetryPolicy.<Void>builder()
      .handle(ArithmeticException.class)
      .handle(IllegalArgumentException.class)
      .waitAndRetryForever((i, r, c) -> Duration.ZERO);

    assertThatExceptionOfType(NullPointerException.class).isThrownBy(() ->
      PolicyUtils.raiseExceptions(policy, 1, i -> new NullPointerException()));
  }

  @Test
  public void shouldThrowWhenSpecifiedExceptionPredicateIsNotSatisfied() {
    final var policy = RetryPolicy.<Void>builder()
      .handle(ArithmeticException.class, e -> false)
      .waitAndRetryForever((i, r, c) -> Duration.ZERO);

    assertThatExceptionOfType(ArithmeticException.class).isThrownBy(() ->
      PolicyUtils.raiseExceptions(policy, 1, i -> new ArithmeticException()));
  }

  @Test
  public void shouldThrowWhenNoneOfTheSpecifiedExceptionPredicatesAreSatisfied() {
    final var policy = RetryPolicy.<Void>builder()
      .handle(ArithmeticException.class, e -> false)
      .handle(IllegalArgumentException.class, e -> false)
      .waitAndRetryForever((i, r, c) -> Duration.ZERO);

    assertThatExceptionOfType(IllegalArgumentException.class).isThrownBy(() ->
      PolicyUtils.raiseExceptions(policy, 1, i -> new IllegalArgumentException()));
  }

  @Test
  public void shouldNotThrowWhenSpecifiedExceptionPredicateIsSatisfied() {
    final var policy = RetryPolicy.<Void>builder()
      .handle(ArithmeticException.class, e -> true)
      .waitAndRetryForever((i, r, c) -> Duration.ofMillis(1));

    PolicyUtils.raiseExceptions(policy, 1, i -> new ArithmeticException());
  }

  @Test
  public void shouldNotThrowWhenOneOfTheSpecifiedExceptionPredicatesAreSatisfied() {
    final var policy = RetryPolicy.<Void>builder()
      .handle(ArithmeticException.class, e -> true)
      .handle(IllegalArgumentException.class, e -> true)
      .waitAndRetryForever((i, r, c) -> Duration.ofMillis(1));

    PolicyUtils.raiseExceptions(policy, 1, i -> new IllegalArgumentException());
  }

  @Test
  public void shouldCallOnRetryOnEachRetryWithTheCurrentException() {
    final var expectedExceptions = List.of("Exception #1", "Exception #2", "Exception #3");
    final var retryExceptions = new ArrayList<Throwable>(3);
    final var policy = RetryPolicy.<Void>builder()
      .handle(ArithmeticException.class)
      .waitAndRetryForever((i, r, c) -> Duration.ZERO,
        fromConsumer4(outcome -> d -> (i, c) -> outcome.onFailure(fromConsumer(retryExceptions::add))));

    PolicyUtils.raiseExceptions(policy, 3, i -> new ArithmeticException("Exception #" + i));

    assertThat(retryExceptions.stream()
      .map(Throwable::getMessage)
      .collect(Collectors.toList()))
      .containsExactlyElementsOf(expectedExceptions);
  }

  @Test
  public void shouldCallOnRetryOnEachRetryWithTheCurrentRetryCount() {
    final var expectedRetryCounts = List.of(1, 2, 3);
    final var retryCounts = new ArrayList<Integer>(3);
    final var policy = RetryPolicy.<Void>builder()
      .handle(ArithmeticException.class)
      .waitAndRetryForever((i, r, c) -> Duration.ZERO,
        fromConsumer4(r -> d -> (retryCount, c) -> retryCounts.add(retryCount)));

    PolicyUtils.raiseExceptions(policy, 3, i -> new ArithmeticException());

    assertThat(retryCounts).containsExactlyElementsOf(expectedRetryCounts);
  }

  @Test
  public void shouldNotCallOnRetryWhenNoRetriesArePerformed() {
    final var onRetryCalled = new AtomicBoolean(false);
    final var policy = RetryPolicy.<Void>builder()
      .handle(ArithmeticException.class)
      .waitAndRetryForever((i, r, c) -> Duration.ofMillis(1),
        fromConsumer4(r -> d -> (i, c) -> onRetryCalled.set(true)));

    assertThatExceptionOfType(IllegalArgumentException.class).isThrownBy(() ->
      PolicyUtils.raiseExceptions(policy, 1, i -> new IllegalArgumentException()));
    assertThat(onRetryCalled.get()).isFalse();
  }

  @Test
  public void shouldCreateNewContextForEachCallToExecute() {
    final var contextValue = new AtomicReference<String>();
    //noinspection ConstantConditions
    final var policy = RetryPolicy.<Void>builder()
      .handle(ArithmeticException.class)
      .waitAndRetryForever((i, r, c) -> Duration.ofMillis(1),
        fromConsumer4(r -> d -> (i, context) -> contextValue.set(context.get("key").toString())));

    PolicyUtils.raiseExceptions(policy,
      Context.of(new HashMap<>() {{
        put("key", "original_value");
      }}),
      1,
      i -> new ArithmeticException());

    assertThat(contextValue.get()).isEqualTo("original_value");

    PolicyUtils.raiseExceptions(policy,
      Context.of(new HashMap<>() {{
        put("key", "new_value");
      }}),
      1,
      i -> new ArithmeticException());

    assertThat(contextValue.get()).isEqualTo("new_value");
  }

  @Test
  public void shouldCalculateRetryDurationsFromCurrentRetryAttemptAndDurationProvider() {
    final var expectedRetryWaits = List.of(Duration.ofMillis(2), Duration.ofMillis(4), Duration.ofMillis(8),
      Duration.ofMillis(16), Duration.ofMillis(32));
    final var actualRetryWaits = new ArrayList<Duration>(5);
    final var policy = RetryPolicy.<Void>builder()
      .handle(ArithmeticException.class)
      .waitAndRetryForever((retryAttempt, r, c) -> Duration.ofMillis((long) Math.pow(2.0, retryAttempt)),
        fromConsumer4(r -> duration -> (i, c) -> actualRetryWaits.add(duration)));

    PolicyUtils.raiseExceptions(policy, 5, i -> new ArithmeticException());

    assertThat(actualRetryWaits).containsExactlyElementsOf(expectedRetryWaits);
  }

  @Test
  public void shouldBeAbleToCalculateRetryDurationsBasedOnTheHandledFault() {
    final var expectedRetryWaits = new HashMap<RuntimeException, Duration>() {{
      put(new ArithmeticException(), Duration.ofMillis(2));
      put(new IllegalArgumentException(), Duration.ofMillis(4));
    }};
    final var actualRetryWaits = new ArrayList<Duration>(2);
    //noinspection SuspiciousMethodCalls
    final var policy = RetryPolicy.<Void>builder()
      .handle(RuntimeException.class)
      .waitAndRetryForever((i, outcome, c) -> outcome.fold(v -> Duration.ZERO, expectedRetryWaits::get),
        fromConsumer4(r -> duration -> (i, c) -> actualRetryWaits.add(duration)));

    final var iterator = expectedRetryWaits.keySet().iterator();
    policy.execute(() -> {
      if (iterator.hasNext()) {
        throw iterator.next();
      }
      return null;
    });

    assertThat(actualRetryWaits).containsExactlyElementsOf(expectedRetryWaits.values());
  }

  @Test
  public void shouldBeAbleToPassRetryDurationFromExecutionToSleepDurationProviderViaContext() {
    final var expectedRetryDuration = Duration.ofMillis(1);
    final var actualRetryDuration = new AtomicReference<Duration>();
    final var defaultRetryAfter = Duration.ofMillis(30);
    final var policy = RetryPolicy.<Void>builder()
      .handle(ArithmeticException.class)
      .waitAndRetryForever((i, r, context) -> context.containsKey("RetryAfter")
          ? (Duration) context.get("RetryAfter")
          : defaultRetryAfter,
        fromConsumer4(r -> duration -> (i, c) -> actualRetryDuration.set(duration)));

    final var failedOnce = new AtomicBoolean(false);
    policy.execute(new HashMap<>() {{
                     put("RetryAfter", defaultRetryAfter);
                   }},
      (Context context) -> {
        context.put("RetryAfter", expectedRetryDuration);
        if (!failedOnce.get()) {
          failedOnce.set(true);
          throw new ArithmeticException();
        }
        return null;
      });

    assertThat(actualRetryDuration.get()).isEqualTo(expectedRetryDuration);
  }
}

File: yarl-core/src/test/kotlin/hr/tjakopan/yarl/retry/WaitAndRetryForeverHandleExceptionTest.kt
package hr.tjakopan.yarl.retry

import hr.tjakopan.yarl.Context
import hr.tjakopan.yarl.Policy
import hr.tjakopan.yarl.test.helpers.raiseExceptions
import org.assertj.core.api.Assertions.assertThat
import java.time.Duration
import kotlin.math.pow
import kotlin.test.Test
import kotlin.test.assertFailsWith

@Suppress("UsePropertyAccessSyntax")
class WaitAndRetryForeverHandleExceptionTest {
  @Test
  fun `should not throw regardless of how many times the specified exception is raised`() {
    val policy = Policy.retry<Unit>()
      .handle(ArithmeticException::class)
      .waitAndRetryForever { _, _, _ -> Duration.ZERO }

    policy.raiseExceptions(3) { ArithmeticException() }
  }

  @Test
  fun `should not throw regardless of how many times one of the specified exception is raised`() {
    val policy = Policy.retry<Unit>()
      .handle(ArithmeticException::class)
      .handle(IllegalArgumentException::class)
      .waitAndRetryForever { _, _, _ -> Duration.ZERO }

    policy.raiseExceptions(3) { IllegalArgumentException() }
  }

  @Test
  fun `should throw when exception thrown is not the specified exception type`() {
    val policy = Policy.retry<Unit>()
      .handle(ArithmeticException::class)
      .waitAndRetryForever { _, _, _ -> Duration.ZERO }

    assertFailsWith(NullPointerException::class) {
      policy.raiseExceptions(1) { NullPointerException() }
    }
  }

  @Test
  fun `should throw when exception thrown is not one of the specified exception types`() {
    val policy = Policy.retry<Unit>()
      .handle(ArithmeticException::class)
      .handle(IllegalArgumentException::class)
      .waitAndRetryForever { _, _, _ -> Duration.ZERO }

    assertFailsWith(NullPointerException::class) {
      policy.raiseExceptions(1) { NullPointerException() }
    }
  }

  @Test
  fun `should throw when specified exception predicate is not satisfied`() {
    val policy = Policy.retry<Unit>()
      .handle(ArithmeticException::class) { false }
      .waitAndRetryForever { _, _, _ -> Duration.ZERO }

    assertFailsWith(ArithmeticException::class) {
      policy.raiseExceptions(1) { ArithmeticException() }
    }
  }

  @Test
  fun `should throw when none of the specified exception predicates are satisfied`() {
    val policy = Policy.retry<Unit>()
      .handle(ArithmeticException::class) { false }
      .handle(IllegalArgumentException::class) { false }
      .waitAndRetryForever { _, _, _ -> Duration.ZERO }

    assertFailsWith(IllegalArgumentException::class) {
      policy.raiseExceptions(1) { IllegalArgumentException() }
    }
  }

  @Test
  fun `should not throw when specified exception predicate is satisfied`() {
    val policy = Policy.retry<Unit>()
      .handle(ArithmeticException::class) { true }
      .waitAndRetryForever { _, _, _ -> Duration.ofMillis(1) }

    policy.raiseExceptions(1) { ArithmeticException() }
  }

  @Test
  fun `should not throw when one of the specified exception predicates are satisfied`() {
    val policy = Policy.retry<Unit>()
      .handle(IllegalArgumentException::class) { true }
      .waitAndRetryForever { _, _, _ -> Duration.ofMillis(1) }

    policy.raiseExceptions(1) { IllegalArgumentException() }
  }

  @Test
  fun `should call on retry on each retry with the current exception`() {
    val expectedExceptions = listOf("Exception #1", "Exception #2", "Exception #3")
    val retryExceptions = mutableListOf<Throwable>()
    val policy = Policy.retry<Unit>()
      .handle(ArithmeticException::class)
      .waitAndRetryForever({ _, _, _ -> Duration.ZERO })
      { outcome, _, _, _ -> outcome.onFailure { retryExceptions.add(it) } }

    policy.raiseExceptions(3) { ArithmeticException("Exception #$it") }

    assertThat(retryExceptions.map { it.message }).containsExactlyElementsOf(expectedExceptions)
  }

  @Test
  fun `should call on retry on each retry with the current retry count`() {
    val expectedRetryCounts = listOf(1, 2, 3)
    val retryCounts = mutableListOf<Int>()
    val policy = Policy.retry<Unit>()
      .handle(ArithmeticException::class)
      .waitAndRetryForever({ _, _, _ -> Duration.ZERO })
      { _, _, retryCount, _ -> retryCounts.add(retryCount) }

    policy.raiseExceptions(3) { ArithmeticException() }

    assertThat(retryCounts).containsExactlyElementsOf(expectedRetryCounts)
  }

  @Test
  fun `should not call on retry when no retries are performed`() {
    var onRetryCalled = false
    val policy = Policy.retry<Unit>()
      .handle(ArithmeticException::class)
      .waitAndRetryForever({ _, _, _ -> Duration.ofMillis(1) })
      { _, _, _, _ -> onRetryCalled = true }

    assertFailsWith(IllegalArgumentException::class) {
      policy.raiseExceptions(1) { IllegalArgumentException() }
    }
    assertThat(onRetryCalled).isFalse()
  }

  @Test
  fun `should create new context for each call to execute`() {
    var contextValue: String? = null
    val policy = Policy.retry<Unit>()
      .handle(ArithmeticException::class)
      .waitAndRetryForever({ _, _, _ -> Duration.ofMillis(1) })
      { _, _, _, context -> contextValue = context["key"].toString() }

    policy.raiseExceptions(Context(mapOf("key" to "original_value")), 1) {
      ArithmeticException()
    }

    assertThat(contextValue).isEqualTo("original_value")

    policy.raiseExceptions(Context(mapOf("key" to "new_value")), 1) {
      ArithmeticException()
    }

    assertThat(contextValue).isEqualTo("new_value")
  }

  @Test
  fun `should calculate retry durations from current retry attempt and duration provider`() {
    val expectedRetryWaits = listOf(
      Duration.ofMillis(2),
      Duration.ofMillis(4),
      Duration.ofMillis(8),
      Duration.ofMillis(16),
      Duration.ofMillis(32)
    )
    val actualRetryWaits = mutableListOf<Duration>()
    val policy = Policy.retry<Unit>()
      .handle(ArithmeticException::class)
      .waitAndRetryForever({ retryAttempt, _, _ ->
        Duration.ofMillis(2.0.pow(retryAttempt).toLong())
      })
      { _, duration, _, _ -> actualRetryWaits.add(duration) }

    policy.raiseExceptions(5) { ArithmeticException() }

    assertThat(actualRetryWaits).containsExactlyElementsOf(expectedRetryWaits)
  }

  @Test
  fun `should be able to calculate retry durations based on the handled fault`() {
    val expectedRetryWaits = mapOf<RuntimeException, Duration>(
      ArithmeticException() to Duration.ofMillis(2),
      IllegalArgumentException() to Duration.ofMillis(4)
    )
    val actualRetryWaits = mutableListOf<Duration>()
    val policy = Policy.retry<Unit>()
      .handle(RuntimeException::class)
      .waitAndRetryForever({ _, outcome, _ ->
        outcome.fold(
          { Duration.ZERO },
          { expectedRetryWaits[it] }) ?: Duration.ZERO
      }) { _, duration, _, _ -> actualRetryWaits.add(duration) }

    val iterator = expectedRetryWaits.iterator()
    policy.execute {
      if (iterator.hasNext()) {
        throw iterator.next().key
      }
    }

    assertThat(actualRetryWaits).containsExactlyElementsOf(expectedRetryWaits.values)
  }

  @Test
  fun `should be able to pass retry duration from execution to sleep duration provider via context`() {
    val expectedRetryDuration = Duration.ofMillis(1)
    var actualRetryDuration: Duration? = null
    val defaultRetryAfter = Duration.ofMillis(30)
    val policy = Policy.retry<Unit>()
      .handle(ArithmeticException::class)
      .waitAndRetryForever({ _, _, context ->
        when (context.containsKey("RetryAfter")) {
          true -> context["RetryAfter"] as Duration
          else -> defaultRetryAfter
        }
      }) { _, duration, _, _ -> actualRetryDuration = duration }

    var failedOnce = false
    policy.execute(mapOf<String, Any>("RetryAfter" to defaultRetryAfter)) { context: Context ->
      context["RetryAfter"] = expectedRetryDuration
      if (!failedOnce) {
        failedOnce = true
        throw ArithmeticException()
      }
    }

    assertThat(actualRetryDuration).isEqualTo(expectedRetryDuration)
  }
}
